### PR TITLE
[extractor/wrestleuniverse] Add extractors

### LINKED ISSUE #6120: Unable to download m3u8, unknown type irvine
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2023.01.06** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

I'm unable to download new videos from the site wrestle-universe.com. Previous videos are able to downloaded using the regular command of a direct link to the m3u8 file, however the video from today is not able to be downloaded. I've tried a couple other programs, as well as adding the cookies and referer flags, but am getting an error about this "unknown url type: irvine"

I downloaded the m3u8 locally and this line is in the file, hopefully it's just something that urllib or yt-dlp needs to learn how to handle

```#EXT-X-KEY:METHOD=AES-128,URI="irvine://aes",IV=0x5BAC71BEB33197D5600337CE86DE7862```

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'https://vod.asset.wrestle-universe.com/9WwMBERet9VdH52yMH3F5/live-a_1080p_2997fps.m3u8', '-o', 's2023e07.mp4']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2023.01.06 [6becd25] (pip)
[debug] Python 3.10.6 (CPython x86_64 64bit) - Linux-5.15.0-58-generic-x86_64-with-glibc2.35 (OpenSSL 3.0.2 15 Mar 2022, glibc 2.35)
[debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2020.06.20, mutagen-1.46.0, secretstorage-3.3.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] Loaded 1760 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2023.01.06, Current version: 2023.01.06
yt-dlp is up to date (2023.01.06)
[generic] Extracting URL: https://vod.asset.wrestle-universe.com/9WwMBERet9VdH52yMH3F5/live-a_1080p_2997fps.m3u8
[generic] live-a_1080p_2997fps: Downloading webpage
[debug] Identified a direct video link
[generic] live-a_1080p_2997fps: Downloading m3u8 information
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] live-a_1080p_2997fps: Downloading 1 format(s): 0
[debug] Invoking hlsnative downloader on "https://vod.asset.wrestle-universe.com/9WwMBERet9VdH52yMH3F5/live-a_1080p_2997fps.m3u8"
[hlsnative] Downloading m3u8 manifest
WARNING: Live HLS streams are not supported by the native downloader. If this is a livestream, please add "--downloader ffmpeg --hls-use-mpegts" to your command
[hlsnative] Total fragments: 3113
[download] Destination: s2023e07.mp4
[download] s2023e07.mp4.part-Frag1 has already been downloaded
[download]   0.0% of ~   4.67GiB at   31.19MiB/s ETA Unknown (frag 1/3113)ERROR: unable to download video data: <urlopen error unknown url type: irvine>
Traceback (most recent call last):
  File "/home/mike/.local/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 3236, in process_info
    success, real_download = self.dl(temp_filename, info_dict)
  File "/home/mike/.local/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 2959, in dl
    return fd.download(name, new_info, subtitle)
  File "/home/mike/.local/lib/python3.10/site-packages/yt_dlp/downloader/common.py", line 444, in download
    ret = self.real_download(filename, info_dict)
  File "/home/mike/.local/lib/python3.10/site-packages/yt_dlp/downloader/hls.py", line 354, in real_download
    return self.download_and_append_fragments(ctx, fragments, info_dict)
  File "/home/mike/.local/lib/python3.10/site-packages/yt_dlp/downloader/fragment.py", line 521, in download_and_append_fragments
    decrypt_fragment(fragment, self._read_fragment(ctx)), fragment['frag_index'], ctx)
  File "/home/mike/.local/lib/python3.10/site-packages/yt_dlp/downloader/fragment.py", line 363, in decrypt_fragment
    decrypt_info['KEY'] = decrypt_info.get('KEY') or _get_key(info_dict.get('_decryption_key_url') or decrypt_info['URI'])
  File "/home/mike/.local/lib/python3.10/site-packages/yt_dlp/downloader/fragment.py", line 353, in _get_key
    _key_cache[url] = self.ydl.urlopen(self._prepare_url(info_dict, url)).read()
  File "/home/mike/.local/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 3730, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib/python3.10/urllib/request.py", line 519, in open
    response = self._open(req, data)
  File "/usr/lib/python3.10/urllib/request.py", line 541, in _open
    return self._call_chain(self.handle_open, 'unknown',
  File "/usr/lib/python3.10/urllib/request.py", line 496, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.10/urllib/request.py", line 1419, in unknown_open
    raise URLError('unknown url type: %s' % type)
urllib.error.URLError: <urlopen error unknown url type: irvine>
```

### PR DESCRIPTION
**Pull request by bashonly**
co-authored by @Grub4K in a major way

Adds support for VOD and PPV content from wrestle-universe.com.

This PR also improves the native HLS downloader so it is capable of accepting an AES-128 key and IV from the info_dict, which is necessitated by this site's m3u8 playlists containing bogus key URIs, with the key being obtained via the site's API instead. `hls_aes_key` and `hls_aes_iv` are the `info_dict`/`formats` fields that can be used to pass the hex string keys and IVs to the native HLS downloader.

The extractor module currently needs to import `Cryptodome` locally in one of its methods; this should be handled with `dependencies` after improvements are made.

Closes #6120

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence): co-authored by @Grub4K

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/wrestleuniverse.py
Comment (pukkandan) [inline yt_dlp/extractor/wrestleuniverse.py:27]:
```suggestion
            WrestleUniverseBaseIE._TOKEN = try_call(lambda: self._get_cookies(
                'https://www.wrestle-universe.com/')['token'].value)
            if not self._TOKEN:
                self.raise_login_required()
```

File: yt_dlp/extractor/wrestleuniverse.py
Comment (pukkandan) [inline yt_dlp/extractor/wrestleuniverse.py:22]: Isn't it simpler to put this in `_real_intitialize`?

File: yt_dlp/extractor/wrestleuniverse.py
Comment (pukkandan) [inline yt_dlp/extractor/wrestleuniverse.py:43]:
```suggestion
            data = json.dumps(data, separators=(',', ':')).encode()
```
is default in py3

File: yt_dlp/extractor/wrestleuniverse.py
Comment (pukkandan) [inline yt_dlp/extractor/wrestleuniverse.py:213]: Imo, it makes sense for this function to return the decrypted key instead of a function

File: yt_dlp/extractor/wrestleuniverse.py
Comment (pukkandan) [inline yt_dlp/extractor/wrestleuniverse.py:200]:
```suggestion
            'title': ('displayName', {str}),
```
Is 'title' expected to be absent? If so, you can add a
```diff
+        'title': '',
         **info
```
before L230

File: yt_dlp/extractor/wrestleuniverse.py
Comment (pukkandan) [inline yt_dlp/extractor/wrestleuniverse.py:205]:
```suggestion
            # FIXME: Bug in traverse_obj
            # 'thumbnails': (('keyVisualUrl', 'alterKeyVisualUrl', 'heroKeyVisualUrl'), {'url': {url_or_none}}),
            'thumbnails': (('keyVisualUrl', 'alterKeyVisualUrl', 'heroKeyVisualUrl'), {url_or_none}, {'url': None}),
```

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/wrestleuniverse.py
Comment (pukkandan) [inline yt_dlp/extractor/wrestleuniverse.py:31]: Why is this saved? It doesn't seem to be used elsewhere

Comment (Grub4K) [reply yt_dlp/extractor/wrestleuniverse.py:31]: It is used each time the `_get_token_cookie` method is called, immediately afterwards (`if self._TOKEN_EXPIRY <= int(time.time())`). We need to check `_TOKEN_EXPIRY` each time we use the token to not send an expired token. If we dont do that inside the method, we would have to duplicate the expiration logic. This is also a reason to not do this in `_real_initialize`.

File: yt_dlp/extractor/wrestleuniverse.py
Comment (Grub4K) [reply yt_dlp/extractor/wrestleuniverse.py:213]: It is cleaner to separate the decryption logic internals from the calling code. The `decrypt` method is also used more than once already. This also encapsulates the respective parts completely, if the en-/decryption changes, only the inner function would need to change. If the api changes, only the outer or call api functions need to change.

File: yt_dlp/extractor/wrestleuniverse.py
Comment (Grub4K) [reply yt_dlp/extractor/wrestleuniverse.py:200]: Should be solved by new `traverse_obj` fixes (`''` is dropped as a value, since its falsy).

File: yt_dlp/extractor/wrestleuniverse.py
Comment (Grub4K) [reply yt_dlp/extractor/wrestleuniverse.py:205]: Same as above, traverse fixes have time to be merged beforehand.

File: yt_dlp/extractor/wrestleuniverse.py
Comment (Grub4K) [reply yt_dlp/extractor/wrestleuniverse.py:23]: See below.

File: yt_dlp/extractor/wrestleuniverse.py
Comment (pukkandan) [inline yt_dlp/extractor/wrestleuniverse.py:129]: Do you have a test case for this?

Comment (bashonly) [reply yt_dlp/extractor/wrestleuniverse.py:129]: I don't have any examples of the API response or nextJS JSON not having a `displayName` field, no. My original thought behind doing this was so that the extractor would always return an info_dict with a `title` field, so that the user could continue to download the video even if the API metadata changed. But I guess it is instead better to raise the error so any problem/change is brought to our attention and can be fixed. I'll remove these

Comment (pukkandan) [reply yt_dlp/extractor/wrestleuniverse.py]: 1. title='' => Auto-generate title
2. title=None => Issue warning and auto-generate title
3. title missing => Raise error

I believe what you want here is (2). Though, we could probably change the behavior of (3) in core code if that'd make things easier. The error for missing title is really just vestigial

Comment (bashonly) [reply yt_dlp/extractor/wrestleuniverse.py]: > we could probably change the behavior of (3) in core code 

I like this idea (but am also OK with 2)

This would suffice, right? 
```diff
diff --git a/yt_dlp/YoutubeDL.py b/yt_dlp/YoutubeDL.py
index fd280726f..676201705 100644
--- a/yt_dlp/YoutubeDL.py
+++ b/yt_dlp/YoutubeDL.py
@@ -2411,10 +2411,7 @@ def check_thumbnails(thumbnails):
     def _fill_common_fields(self, info_dict, final=True):
         # TODO: move sanitization here
         if final:
-            title = info_dict.get('title', NO_DEFAULT)
-            if title is NO_DEFAULT:
-                raise ExtractorError('Missing "title" field in extractor result',
-                                     video_id=info_dict['id'], ie=info_dict['extractor'])
+            title = info_dict.get('title')
             info_dict['fulltitle'] = title
             if not title:
                 if title == '':
```

File: yt_dlp/downloader/hls.py
Comment (pukkandan) [inline yt_dlp/downloader/hls.py:236]:
```suggestion
                            decrypt_info['URI'] = urljoin(man_url, decrypt_info['URI'])
```
The urljoin in utils should handle this

File: yt_dlp/YoutubeDL.py
Comment (pukkandan) [inline yt_dlp/YoutubeDL.py:558]: You can change `_decryption_key_url`

File: yt_dlp/extractor/common.py
Comment (pukkandan) [inline yt_dlp/extractor/common.py:234]: maybe it makes sense to put these inside a dict? like `hls_aes: {key, url, iv}`

Then we can do this to prevent passing such formats to external downloaders
```diff
--- a/yt_dlp/downloader/external.py
+++ b/yt_dlp/downloader/external.py
@@ -104,6 +104,7 @@ def supports(cls, info_dict):
         return all((
             not info_dict.get('to_stdout') or Features.TO_STDOUT in cls.SUPPORTED_FEATURES,
             '+' not in info_dict['protocol'] or Features.MULTIPLE_FORMATS in cls.SUPPORTED_FEATURES,
+            not info_dict.get('aes_key') and not info_dict.get('extra_param_to_segment_url'),
             all(proto in cls.SUPPORTED_PROTOCOLS for proto in info_dict['protocol'].split('+')),
         ))
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/downloader/hls.py
Comment (bashonly) [reply yt_dlp/downloader/hls.py:236]: `urllib.parse.urljoin` is used 2 other places in `hls.py`, should I change those as well?

Comment (pukkandan) [reply yt_dlp/downloader/hls.py:236]: sure

File: yt_dlp/downloader/hls.py
Comment (pukkandan) [inline yt_dlp/downloader/hls.py:160]:
```suggestion
        external_aes_key = traverse_obj(info_dict, ('hls_aes', 'key'))
```
If the extractor sets non string value, it should error

Same below

File: yt_dlp/downloader/hls.py
Comment (pukkandan) [inline yt_dlp/downloader/hls.py:210]: same

File: yt_dlp/downloader/hls.py
Comment (pukkandan) [inline yt_dlp/downloader/hls.py:243]: same

File: yt_dlp/extractor/vzaar.py
Comment (pukkandan) [inline yt_dlp/extractor/vzaar.py:90]:
```suggestion
                    f['hls_aes'] = {'uri': url_templ % ('goose', '') + qs}
```

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/downloader/hls.py
Comment (pukkandan) [inline yt_dlp/downloader/hls.py:164]:
```suggestion
        external_aes_iv = traverse_obj(info_dict, ('hls_aes', 'iv'))
```

File: yt_dlp/extractor/common.py
Comment (bashonly) [reply yt_dlp/extractor/common.py]: I liked the dict idea and went with it. I tweaked the external.py code suggestion to use `traverse_obj` so that it would ignore an `hls_aes` dict with keys but no truthy values. This will eliminate the need for extractor code to *only* return an `hls_aes` dict if it knows for certain there are uri/key/iv values

File: yt_dlp/downloader/hls.py
Comment (bashonly) [reply yt_dlp/downloader/hls.py:185]: ahh mb

File: yt_dlp/extractor/wrestleuniverse.py
Comment (bashonly) [reply yt_dlp/extractor/wrestleuniverse.py:205]: Before this extractor is merged, but after the `traverse_obj` PR is merged, I'd like to commit this change and do some testing. (Apologies if I'm just stating the already-expected course of events)

**Comment by pukkandan** [comment]
Actually, don't merge this. I forgot I wanted to split this into 2 commits. Just let me know once testing is done

**Comment by bashonly** [comment]
> Actually, don't merge this. I forgot I wanted to split this into 2 commits. Just let me know once testing is done

Just finished testing, good to go

**Pull request merged**
